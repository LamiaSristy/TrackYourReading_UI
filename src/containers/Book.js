/* eslint-disable camelcase */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import FormBook from '../components/FormBook';
import { fetchUserBook, createBook, deleteBook } from '../actions/book';
import { loginStatus } from '../actions/user';
import '../stylesheets/Book.css';

class Book extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      addForm: false,
      editForm: false,
      idBook: '0',
    };
  }

  componentDidMount() {
    const { user, fetchUserBook } = this.props;
    const ID = user.user.id;
    fetchUserBook(ID);
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { book } = this.props;
    const {
      addForm, editForm,
    } = this.state;
    return book !== nextProps.book
    || addForm !== nextState.addForm
    || editForm !== nextState.editForm;
  }

  addBook = (name, author, genre, pages) => {
    const { createBook, user } = this.props;
    const { addForm } = this.state;
    const user_id = user.user.id;
    createBook({ name, author, genre, pages, user_id });
    this.setState({
      addForm: !addForm,
    });
  };

  displayForm = () => {
    const { addForm } = this.state;
    this.setState({
      addForm: !addForm,
    });
  }

  displayEdit= e => {
    const { editForm } = this.state;
    this.setState({
      editForm: !editForm,
      idBook: e.target.id,
    });
  }

  deleteBook = id => {
    const { user } = this.props;
    const { deleteBook } = this.props;
    const user_id = user.user.id;
    deleteBook({ user_id, id });
  }

   changeEditForm = () => {
     const { editForm } = this.state;
     this.setState({
       editForm: !editForm,
     });
   }

   changeAddForm = () => {
     const { addForm } = this.state;
     this.setState({
       addForm: !addForm,
     });
   }

   render() {
     const { book } = this.props;
     const {
       addForm, editForm, idBook,
     } = this.state;
     return (
       <main className="main">
         <button type="button" className="add-book" onClick={this.displayForm}>+</button>
         <div className="books">
           { !editForm && !addForm && <h3> All your Books</h3>}

           {book.length === 0 && !addForm && <div className="tracking">Start adding a book you want to track here!</div>}
           {book.map(book => (
             <div key={book.id}>
               { !editForm && !addForm && (
               <div className="one-book">
                 <div className="buttons">
                   <button type="button" onClick={() => this.deleteBook(book.id)}>
                     <i className="fa fa-minus-square" />
                   </button>
                   <button type="button" onClick={this.displayEdit}>
                     <i className="fa fa-pencil-square" id={book.id} />
                   </button>
                 </div>
                 <div className="book-info">
                   <Link to={{
                     pathname: `book/${book.id}`,
                     state: {
                       namebook: book.name,
                     },
                   }}
                   >
                     {!editForm && (
                     <div>
                       <div className="book-name">
                         <p>Name:</p>
                         <p>Author:</p>
                         <p>Genre:</p>
                         <p>Pages:</p>
                       </div>
                       <div className="book-description">
                         <p>{book.name}</p>
                         <p>{book.author}</p>
                         <p>{book.genre}</p>
                         <p>{book.pages}</p>
                       </div>
                     </div>
                     )}
                   </Link>
                 </div>
               </div>
               )}
               { editForm && book.id.toString() === idBook && <FormBook actionToPerform="Save Changes" buttonId={idBook} changeEditForm={this.changeEditForm} />}
             </div>
           ))}
         </div>
         <div className="newbook">
           {addForm && <FormBook addBook={this.addBook} actionToPerform="Add" changeAddForm={this.changeAddForm} />}
         </div>
       </main>
     );
   }
}

const mapStateToProps = state => (
  {
    user: state.user,
    isLogin: state.user.isLogin,
    book: state.book,
  });
const mapDispatchToProps = dispatch => ({
  fetchUserBook: data => dispatch(fetchUserBook(data)),
  createBook: data => dispatch(createBook(data)),
  deleteBook: id => dispatch(deleteBook(id)),
  loginStatus: () => dispatch(loginStatus()),
});

Book.propTypes = {
  fetchUserBook: PropTypes.func,
  createBook: PropTypes.func,
  deleteBook: PropTypes.func,
  user: PropTypes.shape({
    user: PropTypes.shape({
      id: PropTypes.number,
    }),
  }),
  book: PropTypes.arrayOf(PropTypes.shape({
    description: PropTypes.string,
    name: PropTypes.string,
  })),
};

Book.defaultProps = {
  createBook: () => {},
  deleteBook: () => {},
  fetchUserBook: () => {},
  book: {},
  user: {},
};

export default connect(mapStateToProps, mapDispatchToProps)(Book);
