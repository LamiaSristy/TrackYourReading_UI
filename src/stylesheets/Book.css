@import "https://fonts.googleapis.com/css2?family=Raleway:wght@700&display=swap";
@import "https://fonts.googleapis.com/css2?family=Roboto&display=swap";

h3 {
  font-weight: 800;
  color: #164663;
}

.main {
  padding-bottom: 10vh;
  font-family: 'Roboto', sans-serif;
  overflow-y: scroll;
}

.main h3 {
  text-align: center;
  font-family: 'Raleway', sans-serif;
  font-size: 4.5vw;
}

.add-book {
  font-size: 11vw;
  color: white;
  position: absolute;
  right: 11px;
  top: 0;
  background: none;
  border: none;
  cursor: pointer;
}

.books {
  margin-top: 12px;
}

.one-book {
  display: flex;
  justify-content: flex-end;
  border: 1px solid rgb(218, 213, 213);
  background-color: white;
  margin: 0 15px 12px 15px;
  border-radius: 11px;
  flex-direction: column;
  align-items: center;
}

.one-book .buttons {
  align-self: flex-end;
  margin-top: 4px;
}

.one-book .buttons button {
  border: none;
  background: none;
  font-size: 5vw;
  padding: 0 6px;
  cursor: pointer;
}

.fa-minus-square  {
  color: rgb(139, 58, 58);
}

.fa-pencil-square {
  color: black;
}

.one-book:hover {
  background-color: rgb(207, 240, 187, 0.7);
}

.one-book .book-info {
  width: 90%;
  font-size: 3vw;
  margin-bottom: 12px;
}

.one-book .book-info a {
  color: black;
  text-decoration: none;
}

.one-book .book-info a div {
  display:flex;
}

.one-book .book-info p {
  margin: 2px 0;
}

.book-name {
  font-weight: bold;
  align-items: center;
  width: 33%;
  justify-content: flex-start;
}

.book-description {
  width:70%;
  align-items: flex-start;
}

.one-book .book-info .book-name,
.one-book .book-info .book-description {
  display: flex;
  flex-direction: column;
}


.tracking {
  display: flex;
  justify-content: center;
  text-align: center;
  margin-top: 150px;
}
