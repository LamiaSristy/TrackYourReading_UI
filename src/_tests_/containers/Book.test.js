import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Book from '../../containers/Book';

const mockStore = configureMockStore();
const store = mockStore({});
const setup = () => shallow(
  <Provider store={store}>
    <Book />
  </Provider>,
);

describe('Book Container', () => {
  let container;
  beforeEach(() => {
    container = setup();
  });

  test('renders the the page', () => {
    const page = container.find('.main');
    expect(page).toMatchSnapshot();
  });

  test('renders the text of the tracking class', () => {
    const text = container.find('.tracking');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the add-book class', () => {
    const text = container.find('.add-book');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the books class', () => {
    const text = container.find('.books');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the one-book class', () => {
    const text = container.find('.one-book');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the buttons class', () => {
    const text = container.find('.buttons');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the book-info class', () => {
    const text = container.find('.book-info');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the book-name class', () => {
    const text = container.find('.book-name');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the book-description class', () => {
    const text = container.find('.book-description');
    expect(text).toMatchSnapshot();
  });
});
