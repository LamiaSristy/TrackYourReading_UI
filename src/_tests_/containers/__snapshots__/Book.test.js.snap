// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Container renders the text of the add-book class 1`] = `ShallowWrapper {}`;

exports[`Book Container renders the text of the book-description class 1`] = `ShallowWrapper {}`;

exports[`Book Container renders the text of the book-info class 1`] = `ShallowWrapper {}`;

exports[`Book Container renders the text of the book-name class 1`] = `ShallowWrapper {}`;

exports[`Book Container renders the text of the books class 1`] = `ShallowWrapper {}`;

exports[`Book Container renders the text of the buttons class 1`] = `ShallowWrapper {}`;

exports[`Book Container renders the text of the one-book class 1`] = `ShallowWrapper {}`;

exports[`Book Container renders the text of the tracking class 1`] = `ShallowWrapper {}`;

exports[`Book Container renders the the page 1`] = `ShallowWrapper {}`;
